oauth2client/service_account.py

```diff
@@ -12,96 +12,281 @@
+            self._private_key_password)
         self._scopes = util.scopes_to_string(scopes)
+        self._private_key_id = private_key_id
+        self._service_account_id = service_account_id
         self._user_agent = user_agent
-        self._token_uri = token_uri
-        self._revoke_uri = revoke_uri
         self._kwargs = kwargs
 
-    def _generate_assertion(self):
-        """Generate the assertion that will be used in the request."""
+    def _to_json(self, strip, to_serialize=None):
+        """Utility function that creates JSON repr. of a Credentials object.
 
-        header = {
-            'alg': 'RS256',
-            'typ': 'JWT',
-            'kid': self._private_key_id
-        }
+        Over-ride is needed since PKCS#12 keys will not in general be JSON
+        serializable.
+
+        Args:
+            strip: array, An array of names of members to not include in the
+                   JSON.
+            to_serialize: dict, (Optional) The properties for this object
+                          that will serialized. This allows callers to modify
+                          before serializing.
+
+        Returns:
+            string, a JSON representation of this instance, suitable to pass to
+            from_json().
+        """
+        if to_serialize is None:
+            to_serialize = copy.copy(self.__dict__)
+        pkcs12_val = to_serialize.get(_PKCS12_KEY)
+        if pkcs12_val is not None:
+            to_serialize[_PKCS12_KEY] = base64.b64encode(pkcs12_val)
+        return super(ServiceAccountCredentials, self)._to_json(
+            strip, to_serialize=to_serialize)
+
+    @classmethod
+    def from_json_keyfile(cls, filename, scopes=''):
+        """Factory constructor from JSON keyfile.
+
+        Args:
+            filename: string, The location of the keyfile.
+            scopes: List or string, (Optional) Scopes to use when acquiring an
+                    access token.
+
+        Returns:
+            ServiceAccountCredentials, a credentials object created from
+            the keyfile.
+
+        Raises:
+            ValueError, if the credential type is not :data:`SERVICE_ACCOUNT`.
+            KeyError, if one of the expected keys is not present in
+                the keyfile.
+        """
+        with open(filename, 'r') as file_obj:
+            client_credentials = json.load(file_obj)
+
+        creds_type = client_credentials.get('type')
+        if creds_type != SERVICE_ACCOUNT:
+            raise ValueError('Unexpected credentials type', creds_type,
+                             'Expected', SERVICE_ACCOUNT)
+
+        constructor_kwargs = {'scopes': scopes}
+        for key, arg_name in _KEYFILE_MAPPING:
+            if key in client_credentials:
+                constructor_kwargs[arg_name] = client_credentials[key]
+            else:
+                raise KeyError('JSON keyfile missing key', key)
+
+        return cls(**constructor_kwargs)
```

Comment:
Normally I say "always pass positional arguments positionally and keyword arguments by keyword", but this expression looks like it will pass `service_account_email` by keyword. Have I gotten that right? No need to change anything, I'm just making sure I understand.